README.md

```diff
@@ -557,10 +555,10 @@ Staking scripts allow arbitrary logic for the BorrowerIDs such as a multisig.
 ## Future Directions and Considerations
 
 #### Term Extensions/Renegotiations
-A borrower may renegotiate an active loan with their lender, without closing or defaulting on the loan. This may be to "refinance" the loan, to negotiate a loan term extension, or for whatever other reasons the two parties may agree upon. All such actions would be queryable by prospective lenders in the future, giving them further insight into the nature/creditworthiness of the borrower.
+A borrower may renegotiate an active loan with their lender, without closing or defaulting on the loan. This may be to "refinance" the loan, to negotiate a loan term extension, or for whatever other reasons the two parties may agree upon. All such actions would be queryable by prospective lenders in the future, giving them further insight into the borrower's nature/creditworthiness.
 
 #### Multi-Asset Loans
-In addition to using multiple collaterals for loans (which is already implemented), it is possible to create loans where multiple assets are being borrowed. This is especially useful in combination with multi-asset collateral, allowing users to create "packaged" loans that are hedged against the "global" price movements of any one of the underlying assets.
+In addition to using multiple collaterals for loans (which is already implemented), it is possible to create "bundled" loans with multiple borrowed assets. This is especially useful in combination with multi-asset collateral; bundled loans can provide a hedge against fluctuations of any one of the underlying assets.
```

Comment:
I'm not sure how to do this with loan asset beacons. The way the loan asset beacons would work is there would be a unique minting policy for each loan asset that way the asset name can be the loan id. Enabling multi-asset loans with loan asset beacons would require multiple minting policies for a loan like this; each one would have to do basically the same logic (since they can be used in isolation, too). This would kill the composability since it creates redundant executions from these minting policies.

If a general beacon was used instead (like what is currently implemented), there would be no way to know what asset will be paid to the lender just based on the Key NFT. A prospective buyer of Key NFTs would need to query **all** Key NFTs and filter for ones that have the assets they care about. This seems extremely inefficient and likely impractical.

Considering this protocol was designed for p2p, which usually implies individuals or small groups, I think loan asset beacons are more important than hedging, which is usually for large corporations. (This judgement is based on how I would use the protocols.) A dedicated protocol can be created for large corporations which can support hedging like this. The difficulty with using beacons may be a non-issue for that separate protocol since the context is different.

Reply:
I think the above comment over-complicates the loan asset beacon. A simpler approach may be to simply add another beacon to the current implementation. The beacon would be the loan asset beacon where the name is:
```
sha256( loan_asset_policy_id ++ loan_asset_token_name )
```

The Ask phase would require this new beacon to be minted and stored with the `Ask` beacon. The loan asset beacon would carry over into the Active phase. This approach allows for the following queries:
1. all active loans => query just the `Active` beacons
2. all asks => query just the `Ask` beacons
3. all offers => query just the 'Offer' beacons
4. all active loans for asset X => query UTxOs with both an `Active` beacon and the loan asset beacon for asset X
5. all asks for asset X => query UTxOs with both an `Ask` beacon and the loan asset beacon for asset X
6. key/lock NFT queries
7. BorrowerId and LenderId for all loans.

If another loan asset beacon is minted for the Offer phase, then it would also be possible to query all Offer UTxOs for asset X.

A drawback of this approach is that the Key NFT still does not convey which asset the loan is for so the secondary market buyers would need to query all Key NFTs for sale just to filter for a specific loan asset. However, if buyers know what loan assets they are interested in, they can query the Key NFTs + the loan asset beacon for that asset. If no utxo exists with both those assets, it can just return an empty list. Perhaps this is enough of a filter? I think this would be faster/cheaper than parsing the datums for every Key NFT and filtering out based on info in the datum.

For multi-asset loans, there could just be additional loan asset beacons minted (and the datum could contain a list of loan assets). However, there are then a few other issues/questions:
1. If 100 ADA and 10 AGIX are borrowed, does repayment only care about relative value (ie the loan can be fully paid off using just ADA) or do the specific assets need to be repaid (ie 100 ADA  and 10 AGIX)?
2. The collateralization field for the datum would grow by `n * m` where `n` is the number of loan assets and `m` is the number of assets used for collateral. There would need to be checks to make sure all combinations are accounted for. I think the extra logic required for these checks and calculations might be quite substantial.

There may be room for multi-asset loans but I would like to see how much room is left after all the other features are added.